### PR TITLE
Improved `readInt` implementation

### PR DESCRIPTION
**Pull request by vdukhovni**
- Removes arbitrary 18 byte limit on decimal input.

  * This former limit means that long decimal strings could on repeated
    be incorrectly parsed as multiple ints.
  * Also 18 bytes is not even enough for 64-bit minBound or maxBound.

- Significantly speeds up readInt by avoiding "toStrict", and handling
  chunk boundaries efficiently.

- Adds a `readInt'` function that supports transformation of the stream
  tail to prepare for reading more input, by e.g.  dropping leading
  whitespace while we are still holding the raw chunk, avoiding the cost
  of burying it deep in a closure, only to later find it again and trim.

- Reading ~100M space separated ints now takes ~5.5s vs. over 25s with
  the original implementation.

- This does however mean that on an unbounded stream of digits the new
  `readInt` is not guaranteed to ever return.  The documentation notes
  that users can elect to split the stream and only run `readInt` on
  some bounded initial segment.

REVIEW topics:

If someone wants to bikeshed a better name for the `Chunker` type, I
am not particularly fixed on the current name, though it seems OK to
me.  Or we could drop the type alias and use the full expansion in
the signature of `readInt`', but that could make the haddock
difficult to fit horizontally on the screen.

Note that `finiteBitSize` requires base 4.7, i.e. at least GHC 7.8.  The CI
does not test anything older than 7.10, so I assume that 7.6 is no longer
supported.  Otherwise a work-around is needed for older versions.

The `LambdaCase` and `MultiWayIf` pragmas are also used, these
require at least GHC 7.6.

Cc: @bodigrim, @chessai

### CODE REVIEW

**Comment by fosskers** [comment]
The lowest GHC we support is 7.10. Thanks also for adding all those test cases!

File: Data/ByteString/Streaming/Char8.hs [hunk 1]
```diff
@@ -710,20 +713,155 @@ putStrLn :: MonadIO m => ByteString m r -> m r
...
+        -> m (Compose (Of (Maybe Int)) (ByteString m) r)
+{-# INLINE readInt #-}
+readInt = readInt' Chunk
+
+-- | Try to read an 'Int' value from the 'ByteString', returning 'm
+-- (Compose (Just n :> t))' on success, where @n@ is the value read and
+-- @t@ is the rest of the input streaming 'ByteString' computed by
+-- applying the provided 'Chunker' function to the first non-empty chunk
+-- following the parsed number and the rest of the stream.  The function
+-- is not called, and the stream is returned as-is, preceding by
```
Comment (vdukhovni) [inline Data/ByteString/Streaming/Char8.hs]: I'll fix the typo here...
```suggestion
-- is not called, and the stream is returned as-is, preceded by
```

File: Data/ByteString/Streaming/Char8.hs [hunk 2]
```diff
@@ -710,20 +713,155 @@ putStrLn :: MonadIO m => ByteString m r -> m r
...
+-- whitespace, positioning the stream at the next potential input value.
+-- It is up to the function to discard whitespace from any subsequent
+-- chunks if the remainder of the first chunk is all whitespace, it is
+-- not called recursively on its own output even when the result would
+-- have an empty first chunk.
+--
+-- If the stream of digits decodes to a value larger than can be
+-- represented by an 'Int', the result will overflow, and will not match
+-- the input digits.  Input is consumed until no more digits remain.  If
+-- the stream is an unbound run of decimal digits, this function may
```
Comment (vdukhovni) [inline Data/ByteString/Streaming/Char8.hs]: Ditto.
```suggestion
-- the stream is an unbounded run of decimal digits, this function may
```

File: Data/ByteString/Streaming/Char8.hs [hunk 3]
```diff
@@ -710,20 +713,155 @@ putStrLn :: MonadIO m => ByteString m r -> m r
...
+-- never return.  If that's a concern, you can split the stream into an
+-- initial segment with an apporpriate maximum length and a tail, and
```
Comment (vdukhovni) [inline Data/ByteString/Streaming/Char8.hs]:
```suggestion
-- initial segment with an appropriate maximum length and a tail, and
```

File: Data/ByteString/Streaming/Char8.hs [hunk 4]
```diff
@@ -710,20 +713,155 @@ putStrLn :: MonadIO m => ByteString m r -> m r
...
+-- apply 'readInt' to just the initial segment.
```
Comment (vdukhovni) [inline Data/ByteString/Streaming/Char8.hs]:
```suggestion
-- apply `readInt'` to just the initial segment.
```

File: Data/ByteString/Streaming/Char8.hs [hunk 5]
```diff
@@ -710,20 +713,155 @@ putStrLn :: MonadIO m => ByteString m r -> m r
...
+readInt' :: Monad m
+         => Chunker m r    -- ^ 'Chunker' function for the tail of the stream
+         -> ByteString m r -- ^ Input stream
+         -> m (Compose (Of (Maybe Int)) (ByteString m) r)
+{-# INLINE readInt' #-}
+-- While dosigned, doplus and dominus are recursive and won't inline, it
+-- pays to inline `readInt'`, because the body of the 'Chunker' can be
+-- inlined into the recursive functions, substantially improving the
+-- performance, e.g. when it is just 'Chunk' as in `readInt,` or when it
+-- is function to skip trailing whitespace after the input (leading
```
Comment (vdukhovni) [inline Data/ByteString/Streaming/Char8.hs]:
```suggestion
-- is a function to skip trailing whitespace after the input (leading
```

File: Data/ByteString/Streaming/Char8.hs [hunk 6]
```diff
@@ -710,20 +713,155 @@ putStrLn :: MonadIO m => ByteString m r -> m r
...
+-- whitespace in the tail) preparing the stream to read the next value.
+readInt' chunker = dosigned
+  where
+    done !n !s = pure $ Compose $ Just n  :> s
+    nada !s    = pure $ Compose $ Nothing :> s
+    {-# INLINE done #-}
+    {-# INLINE nada #-}
+
+    -- Optimise for the case where the desired integer is present, and
+    -- the first chunk is not empty or has just the sign, in which case
```
Comment (vdukhovni) [inline Data/ByteString/Streaming/Char8.hs]:
```suggestion
    -- the first chunk is neither empty nor has just the sign, in which case
```

File: Data/ByteString/Streaming/Char8.hs [hunk 7]
```diff
@@ -710,20 +713,155 @@ putStrLn :: MonadIO m => ByteString m r -> m r
...
+                           | w == 0x2d     -> dominus 0 cs
+                           | otherwise     -> nada bs
+                   | otherwise -> nada bs
+            | otherwise -> dosigned cs
+        Go m -> m >>= dosigned
+        e@(Empty _) -> nada e
+
+    -- Finish reading a positive value, given the value of its prefix.
+    -- For a 64-bit 'Int', only the last 64 digits matter.  Thus, if the
+    -- new chunk has more than 64-bytes of initial digits, we don't need
```
Comment (vdukhovni) [inline Data/ByteString/Streaming/Char8.hs]:
```suggestion
    -- new chunk has more than 64 digits, we don't need
```

File: Data/ByteString/Streaming/Char8.hs [hunk 8]
```diff
@@ -710,20 +713,155 @@ putStrLn :: MonadIO m => ByteString m r -> m r
...
+    -- to combine the result with anything from the previous chunk.
+    --
+    -- But we still pay the cost of multiply by 10 and add for each
+    -- digit, and typically the number of digits is small, so testing
+    -- for long runs and using only the last 64 or so digits is likely
+    -- slower in in practice for anything but unreasonably long digit
```
Comment (vdukhovni) [inline Data/ByteString/Streaming/Char8.hs]:
```suggestion
    -- slower in practice for anything but unreasonably long digit
```

**Review by vdukhovni — COMMENTED** [review]
Noting doc typos for a fixup commit.

**Comment by vdukhovni** [comment]
Please don't merge this, it does not take effects into account correctly. I'll rework it to address the issue.

[ Actually, perhaps not a problem, I was thrown off by thinking about how overflow handling might work, but since we're not doing overflow detection, I think this is OK, but just in case, I'll review again... ]

**Comment by vdukhovni** [comment]
I pushed a new commit that handles cases where the leading `+` or `-` is at a chunk boundary, and what follows is not an unsigned integer. This needs to return `Nothing` not `0`, but also return a leftover string with the `+` or `-` sign as a short chunk in front of the chunk with the non-integer payload.  Since it is not too difficult (but quite wrong) to accidentally return just the original input stream (one of whose _effects_ is already performed), I enhanced the tests to verify that we're not performing too few or too many effects by the time the entire stream has been processed.

Also added tests to check some more  "+"/"-" at chunk boundary cases.

**Comment by vdukhovni** [comment]
I am prototyping (no PR yet), a new implementation that restores (more correct) overflow protection:

* All valid inputs are correctly converted, even those with 19 digits (or sign + 19 digits), supporting the full range from `minBound` to `maxBound`
* All inputs that would overflow return `Nothing` and the invalid input as the tail of the string, but internally, that input is magically reconstructed from the accumulated partial number (last value prior to would-be overflow) and the number of digits consumed, by generating the corresponding decimal string with enough leading zeros and the original explicit sign (if any). This frees us from having to copy the input.
* To avoid being stuck forever reading infinite streams of zeros, a safety mechanism gives up after reading ~32k leading zeros, at some point the input is no longer valid, but rather an attack. The logic is not go to the next chunk if we've already accumulated that many bytes.  However, if the chunk we already have in memory is for some reason already large, then I'm willing to parse it through to the end.
* Accompanying this, I will add a variant that automatically skips trailing whitespace after parsing each number, which makes reading streams of numbers separated by whitespace much more efficient. This too has a safely limit, after ~32k of trailing whitespace, no more whitespace is skipped, again to avoid being stuck forever on hostile inputs, if reading from a network or other unbound source.

The performance is only slightly worse than this PR, for 100M space-separated Ints that are 1..100M in random order, the time to read and add them all up changes from ~4.84s to 4.88s on my (somewhat dated) Intel CPU.

By introducing the "tail trimming" variant of `readInt`, I'm inclined to drop the generic "Chunker" interface in this PR as being too fancy. In practice users will likely want just the verbatim remainder, or the trimmed version.  The only thing they won't get that way is the ability to control the safety mechanism, by e.g. choosing a different limit on the number of spaces consumed, or setting no limit at all. [ Edit: I'll look into whether good performance is preserved if the whitespace skip is instead performed before reading each integer, rather than after, this is more friendly for users who intermix reading integers with other consumers of stream data, but if this is done, the whitespace will be dropped unconditionally, even when ultimately returning `Nothing`. ]

So before I open a "competing" PR, I'd like some feedback on what folks think about this version versus the description above?

I should also mention that while the code could In principle handle other bounded integral types, the approach can only work for decimal (or octal) inputs, with hex the input can use mixed case, and I can no longer reconstruct the input from the partly accumulated digits, which may have originated from separate chunks, making the code more complex and likely much slower.

Internally the accumulator is a `Word` value, and all the digits are combined unsigned into the word, what's variable is the bounds at which I detect overflow. So handling `Int32` or `Word32`, ... would just be a matter of tweaking the bounds, but I am not inclined to address that at this time.  Support for the other types was not previously available.  We can look into that later.

One of the complications of handling the various (Bounded) integral types is that, unlike `C`, Haskell has no `uintmax_t` equivalent.  The closest we've got is the `Word64` and `Int64` types, and there's some complex logic around how these have been bolted on in 32-bit vs. 64-bit systems.  It seems that's starting to change, IIRC a recent PRs was introducing `Word64#` primitives unconditionally.

Anyway, more could be done, but this is what I have now...  Time for feedback...

CC: @chessai, @hsyl20, @bodigrim, @sjakobi, @cartazio

**Comment by Bodigrim** [comment]
> Internally the accumulator is a Word value, and all the digits are combined unsigned into the word, what's variable is the bounds at which I detect overflow. 

BTW there is [`timesWord2`](http://hackage.haskell.org/package/ghc-prim-0.6.1/docs/GHC-Prim.html#v:timesWord2-35-), which may be handy to detect overflows.

**Comment by vdukhovni** [comment]
> > Internally the accumulator is a Word value, and all the digits are combined unsigned into the word, what's variable is the bounds at which I detect overflow.
> 
> BTW there is [`timesWord2`](http://hackage.haskell.org/package/ghc-prim-0.6.1/docs/GHC-Prim.html#v:timesWord2-35-), which may be handy to detect overflows.

Yes, I was aware of it, but since I have to a comparison either way (before multiplication or after), I went with the simpler "before" option. If the accumulator is below 1/10 of the maximum representable value, I keep going, if above I fail. And if equal, I check the input digit to see whether it is below or above the residue of the maximum value mod 10.  This runs quite efficiently, and for most numbers only one test is needed, which I think is no more expensive than using and testing the result of `timesWord2#`.

My experience with using numeric primops is that it is easy (and somewhat counter-intuitive) to actually get worse performance with them, because GHC seems to do fewer optimisations on code with primops, or because it is able to figure out a better set of primops to apply than I can, but many times when I thought I could write better code directly (avoiding allocations, ...) I found that GHC typically already used unlifted values wherever possible most of the time, and seems to have produced better code in most cases. :-)

There are a few cases in which I was able to get better code with priomops, but without knowing more about internals than I do, I struggle to intuit which are the cases where it is likely to yield good results.

But while, I have your attention, direct feedback on the substance of this PR, and proposed alternative would be very much appreciated.  The tangential comments are of course also welcome! :-)
